### DISCUSSION TITLE
Shared memory and numpy arrays

### DISCUSSION THREAD

**Discussion by PaulRudin** [v6d-io/v6d#1329] [Q&A] [answered]
Greetings once again!

```
>>> a = c.get("o000066a0a7c6fb1c")
>>> a
ndarray([      14,      226,      276, ...,   131315,  3220012, 11171032],
        dtype=int32)
>>> b = c.get("o000066a0a7c6fb1c")
>>> a is b
False
>>> a.data is b.data
False
>>> a.data
<memory at 0x7f943d167340>
>>> b.data
<memory at 0x7f943d167700>
```
I think I was expecting the underlying data to be the same bit of memory. I guess that's not so?

**Comment by PaulRudin** [comment]
Ah - maybe it is?
```
>>> a.base.data
<memory at 0x7f943d167340>
>>> b.base.data
<memory at 0x7f943d167640>
```

**Reply by PaulRudin** [reply]
Ah, those are not the same... just close.

**Reply by sighingnow** [reply]
> Ah, those are not the same... just close.

`a.base.data` and `b.base.data` is not the memory address.

**Reply by sighingnow** [reply]
Try run `a.base.data` multiple times, you'll see different output for each run.

**Comment by sighingnow** [comment, marked as answer]
@PaulRudin  `numpy.ndarray.base.data` is not the memory address of the array payload. With vineyard, get the same object twice **with the same client instance**, the memory address should be the same.

```python
In [3]: import vineyard
import
In [4]: import numpy as np

In [5]: client = vineyard.connect("/tmp/vineyard.sock")

In [6]: object_id = client.put(np.ones((10000, 1000)))

In [7]: r1 = client.get(object_id)

In [8]: r2 = client.get(object_id)

In [9]: r1 is r2
Out[9]: False

In [10]: r1.__array_interface__["data"]
Out[10]: (140492393889856, True)

In [11]: r2.__array_interface__["data"]
Out[11]: (140492393889856, True)

In [12]: r1.__array_interface__["data"] == r2.__array_interface__["data"]
Out[12]: True
```

See https://stackoverflow.com/questions/11264838/how-to-get-the-memory-address-of-a-numpy-array-for-c for more discussion about the memory address of numpy ndarray.
